feat: add utility to generate integration common names automatically

saikumarrs opened the pull request:
## Description of the change

Added a utility to auto-generate the generic common names for an integration given its display name. For integrations where this default common names map is not enough, we can still define additional entries in the `CNameMapping` object in `constants.js`.

The default common name mapping is built by converting the display name to lowercase. Hence, when comparing any name from the integrations object, it is also first converted to lowercase.

Additionally, refactored the `IsEventBlackListed` function to remove redundant common names resolution logic.

This feature has two benefits:
- Reduced SDK size
- Less error prone w.r.t common names map definition.

It is currently implemented as an example for the "Adobe Analytics" destination.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

## Checklists

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/rudderlabs/rudder-sdk-js/685)
<!-- Reviewable:end -->

bardisg commented on utils/utils.js:
```diff
@@ -286,12 +286,17 @@ function getRevenue(properties, eventName) {
     if (integrationObject.hasOwnProperty(key)) {
-      if (commonNames[key]) {
-        integrationObject[commonNames[key]] = integrationObject[key];
+      const sanitizedIntgName = key.toLowerCase().trim();
```
Not sure about this one, as I do not know the permutations of the key values.

It seems like only the display name is cast to lowercase in the utility method, for example NAME or any hardcoded CNameMapping coould have uppercase values.

If the key value in this context is alwys the display name this is fine elsewise toLowerCase would not work for all cases

saikumarrs replied:
Yes, the keys in the common names map are always a combination of the display name. 

The casing is made to lower on both sides to eliminate any play w.r.t that. One would ideally override (rather, add new ones) the cname mapping for other combinations (ex: "Customer.io" could have other hardcoded names defined such as "customer-io", "kustomerio" (with a "K"), etc).

However, I've now fixed the bug (converted the case of even the hard-coded cname mappings)

bardisg replied:
got it. LGTM

bardisg commented on utils/utils.js:
```diff
@@ -286,12 +286,17 @@ function getRevenue(properties, eventName) {
-        if (commonNames[key] != undefined && commonNames[key] != key) {
+        if (
+          commonNames[sanitizedIntgName] != undefined &&
+          commonNames[sanitizedIntgName] != key
```
we should be using !==

bardisg replied:
I see

MoumitaM commented on integrations/utils/commonUtils.js:
```diff
@@ -163,6 +163,40 @@ function getDestinationExternalID(message, type) {
+    // match the existing common names
+    Object.keys(CNameMapping).forEach((name) => {
+      const sanitizedName = name.toLowerCase().trim();
+      cNamesMap[sanitizedName] = CNameMapping[name];
```
`cNamesMap[sanitizedName] = CNameMapping[name];` this will create duplicate records as we are not removing the existing key-value. Is this expected?

saikumarrs replied:
It'll override if a key is already present.

MoumitaM reviewed (commented):
With this change developers now have to keep in mind what combination will be generated by the SDK and also they have to add common names that don't fall into this category. Previously they used to write the all possible combination by themselves. Just thinking out loud, wouldn't this create more confusion and lead to more mistakes?

saikumarrs commented:
> With this change developers now have to keep in mind what combination will be generated by the SDK and also they have to add common names that don't fall into this category. Previously they used to write the all possible combination by themselves. Just thinking out loud, wouldn't this create more confusion and lead to more mistakes?

This utility should tackle more than 90% (100% in some cases) of the common names for all the destinations. I believe we can add documentation to clarify the combinations that are taken care of automatically. Even if someone makes a mistake, they'd just redundantly override a common name.

MoumitaM commented:
@saikumarrs , We can not merge this feature. 
`tranformToRudderNames` is a common function to transform user provided name. But we have modified it. So the existing implementation might break. 
Ex:
```
{
   integrations: {
       All: false,
       "Amplitude": true
   }
}
```
Now there is no mapping b/w "amplitude" to "AM" in common names and modified integrations object will become:
```
{
   integrations: {
       All: false,
       "Amplitude": true
   }
}
```
So no event will reach `Amplitude`.

saikumarrs commented:
> @saikumarrs , We can not merge this feature. `tranformToRudderNames` is a common function to transform user provided name. But we have modified it. So the existing implementation might break. Ex:
> 
> ```
> {
>    integrations: {
>        All: false,
>        "Amplitude": true
>    }
> }
> ```
> 
> Now there is no mapping b/w "amplitude" to "AM" in common names and modified integrations object will become:
> 
> ```
> {
>    integrations: {
>        All: false,
>        "Amplitude": true
>    }
> }
> ```
> 
> So no event will reach `Amplitude`.

Really good catch!
I've fixed this issue. Please review.

MoumitaM reviewed (changes requested):
We need to update the `IsEventBlackListed` function.
```
IsEventBlackListed(eventName, intgName) {
    if (!eventName || !(typeof eventName === "string")) {
      return false;
    }
    const sdkIntgName = commonNames[intgName];
    const intg = this.clientIntegrations.find(
      (intg) => intg.name === sdkIntgName
    );
```
This `intgName` is the name of an integration mentioned inside the constructor fn. 
now there is no mapping b/w ADOBE_ANALYTICS -> ADOBE_ANALYTICS. So this feature will not work for Adobe.
We need to update the logic with:
```
const sdkIntgName = commonNames[intgName] || commonNames[intgName.toLowerCase().trim()];
```

saikumarrs commented:
> 

It's good that you brought it up. We don't need the common name resolution at all, as the names in the destination instance object and the integrations object are already in the rudder names format.
Fixed it.

saikumarrs commented:
> We need to update the `IsEventBlackListed` function.
> 
> ```
> IsEventBlackListed(eventName, intgName) {
>     if (!eventName || !(typeof eventName === "string")) {
>       return false;
>     }
>     const sdkIntgName = commonNames[intgName];
>     const intg = this.clientIntegrations.find(
>       (intg) => intg.name === sdkIntgName
>     );
> ```
> 
> This `intgName` is the name of an integration mentioned inside the constructor fn. now there is no mapping b/w ADOBE_ANALYTICS -> ADOBE_ANALYTICS. So this feature will not work for Adobe. We need to update the logic with:
> 
> ```
> const sdkIntgName = commonNames[intgName] || commonNames[intgName.toLowerCase().trim()];
> ```


It's good that you brought it up. We don't need the common name resolution at all, as the names in the destination instance object and the integrations object are already in the rudder names format.
Fixed it.

MoumitaM commented on integrations/integration_cname.js:
```diff
@@ -135,6 +136,7 @@ const commonNames = {
   ...Refiner,
   ...YandexMetrica,
   ...Qualaroo,
+  all: "All",
```
```suggestion
  All: "All",
```

saikumarrs replied:
The fix is applied in a different location.

MoumitaM reviewed (changes requested):
Found another bug, `All: true/false` was getting converted to `All: undefined` and all the destinations started loading.
